### ISSUE TITLE
Test failure at -O0

### ISSUE THREAD

**Issue by quasicomputational** [haskell/text#236]
```
$ cabal new-test all --disable-optimisation
[...]
    t_compareLength: [Failed]
*** Failed! Falsifiable (after 4 tests and 1 shrink): 
""
-1
(used seed 1847101756120191161)
```

This is with HEAD at 77da9f6a944284b5bb00a60905137dc1b37732ae.

**Comment by hvr** [comment]
@quasicomputational when you say "HEAD", are you referring to GHC HEAD?

**Comment by quasicomputational** [comment]
Ah, no, I mean `text`. I think that was with GHC 8.4.3.

**Comment by hvr** [comment]
@quasicomputational it'd be worth retrying w/ 8.4.4 as well as other GHC versions to see how this is dependent on GHC versions

**Comment by jinwoo** [comment]
We're seeing this failure with `text` 1.2.3.1 and GHC 8.6.4.

**Comment by Javran** [comment]
Investigated a bit with @jinwoo  on this today:
As OP indicated, the failure happens when calling `compareLength "" (-1)`.

In this example [compareLength](https://github.com/haskell/text/blob/2176eb38b5238e763e8076b0d0db8c2f2014ab8b/Data/Text.hs#L640) calls `codePointsSize` with `-1`: https://github.com/haskell/text/blob/2176eb38b5238e763e8076b0d0db8c2f2014ab8b/Data/Text/Internal/Fusion/Common.hs#L328

However `codePointsSize` assumes that input is non-negative: https://github.com/haskell/text/blob/2176eb38b5238e763e8076b0d0db8c2f2014ab8b/Data/Text/Internal/Fusion/Size.hs#L63-L69

This is done using an assertion, which is only enabled when `ASSERTS` is set (this is true for the test suite and only true for this library in dev mode). This assertion error is what we believe to be the source of falsifiability.

Luckily this doesn't practically cause too much trouble in terms of correctness: following along the definition of `compareSize`, it is true that `compareSize len n'` always return `GT` when `n'` is constructed from a negative number and `n' /= (minBound :: Int)`, since len is the length of some Text, which is always >= 0.

**Comment by Lysxia** [comment]
This was fixed in 35d53cd (PR #335).
